sound/soc/sof/topology.c

```diff
@@ -895,24 +895,46 @@ static int sof_parse_string_tokens(struct snd_soc_component *scomp,
 			if (tokens[j].token != le32_to_cpu(elem->token))
 				continue;
 
-			/* matched - now load token */
-			tokens[j].get_token(elem, object,
-					    offset + tokens[j].offset,
-					    tokens[j].size);
+			/* load token if object is not NULL */
+			if (object) {
+				tokens[j].get_token(elem, object, offset + tokens[j].offset,
+						     tokens[j].size);
+				continue;
+			}
+
+			if (!tuples && !num_tuples)
+				continue;
 
+			/* save tuples in the tuple array otherwise */
+			if (!(*num_tuples)) {
+				new_tuples = devm_kzalloc(scomp->dev, sizeof(struct snd_sof_tuple),
+							  GFP_KERNEL);
+				if (!new_tuples)
+					return -ENOMEM;
+			} else {
+				new_tuples = devm_krealloc(scomp->dev, *tuples,
+							   sizeof(struct snd_sof_tuple) *
+							   (*num_tuples + 1), GFP_KERNEL);
```

Comment:
is realloc will invoked a few times in this function ?   if yes, can we do it in two pass ?  calculate total count in first pass only and allocate in the second pass

Reply:
But why? That's not optimal. It's better to parse and add as we go along. The tuple array can be quite large in practice

Reply:
I don't like realloc as I also tried this function and some one told me reviewers don't like realloc

Reply:
we cant prevent realloc because we have multiple token arrays to be parsed per object. The function is going to become unnecessary clumsy if I have the same loop twice.

Reply:
@RanderWang i've removed the krealloc now. hope you're happy now